[FSDP] Provide a utility API to allow users easily to set state_dict_type

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #73716

fsdp_state_dict_type() traverses the modules and calls state_dict_type when it is necessary, eliminating the burden from the users.

Differential Revision: [D34532321](https://our.internmc.facebook.com/intern/diff/D34532321/)

fegin committed:
[FSDP] Provide a utility API to allow users easily to set state_dict_type

fsdp_state_dict_type() traverses the modules and calls state_dict_type when it is necessary, eliminating the burden from the users.

Differential Revision: [D34532321](https://our.internmc.facebook.com/intern/diff/D34532321/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Provide a utility API to allow users easily to set state_dict_type"


fsdp_state_dict_type() traverses the modules and calls state_dict_type when it is necessary, eliminating the burden from the users.

Differential Revision: [D34532321](https://our.internmc.facebook.com/intern/diff/D34532321/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Provide a utility API to allow users easily to set state_dict_type"


fsdp_state_dict_type() traverses the modules and calls state_dict_type when it is necessary, eliminating the burden from the users.

Differential Revision: [D34532321](https://our.internmc.facebook.com/intern/diff/D34532321/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Provide a utility API to allow users easily to set state_dict_type"


fsdp_state_dict_type() traverses the modules and calls state_dict_type when it is necessary, eliminating the burden from the users.

Differential Revision: [D34532321](https://our.internmc.facebook.com/intern/diff/D34532321/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP] Provide a utility API to allow users easily to set state_dict_type"


fsdp_state_dict_type() traverses the modules and calls state_dict_type when it is necessary, eliminating the burden from the users.

Differential Revision: [D34532321](https://our.internmc.facebook.com/intern/diff/D34532321/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/fully_sharded_data_parallel.py:
```diff
@@ -760,12 +760,15 @@ def _need_prefetch_post_backward_hook(self) -> bool:
+    @staticmethod
     @contextlib.contextmanager
-    def state_dict_type(self, state_dict_type: StateDictType) -> Generator:
+    def state_dict_type(module: nn.Module, state_dict_type: StateDictType) -> Generator:
```
maybe rename it to `fsdp_state_dict_type`, unless you feel that the call `FSDP.state_dict_type()` is already clear enough that it's setting things for only FSDP modules.

fegin replied:
I feel `FSDP.state_dict_type()` explicit enough. Only FSDP internal code will use something like `module.state_dict_type()` or `self.state_dict_type()`.

rohan-varma commented on test/distributed/fsdp/test_fsdp_state_dict.py:
```diff
@@ -197,7 +197,7 @@ def _load_state_dict(
             raise ValueError(f"No state_dict for {state_dict_type}")
 
-        with model.state_dict_type(enum_val):
+        with model.state_dict_type(model, enum_val):
```
why not `FSDP.state_dict_type`?

fegin replied:
Good catch!

rohan-varma reviewed (approved):
LGTM, just some minor comments and questions

fegin committed:
Update on "[FSDP] Provide a utility API to allow users easily to set state_dict_type"


fsdp_state_dict_type() traverses the modules and calls state_dict_type when it is necessary, eliminating the burden from the users.

Differential Revision: [D34532321](https://our.internmc.facebook.com/intern/diff/D34532321/)

[ghstack-poisoned]
